File: syntax/Python/backtracking/backtracking.py

def display(a):
    l = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K']
    print([l[a[i]] for i in range(len(a))])

    
def backtracking1(n, m, check, handle):
    def dfs(a, k):      
        for i in range(n):
            a[k] = i
            if(check(a, k)):
                if(k == m - 1):
                    handle(a)
                else:
                    dfs(a, k + 1)
    a = [0 for i in range(m)]
    dfs(a, 0)

def backtracking(n, m, check, handle):
    k, a = 0, [-1 for i in range(m)]
    while(k >= 0):
        a[k] += 1
        while(a[k] < n and not check(a, k)):
            a[k] += 1
        if(a[k] == n or k == m):
            k -= 1
        else:
            if(k == m - 1):
                handle(a)
            else:
                k += 1
                a[k] = -1

def counter(n, m):
    backtracking(n, m, lambda a, k : True,\
                 lambda a : print(a))

#counter(10, 3)

def permutation(n, m):
    def check(a, k):
        for i in range(k):
            if(a[i] == a[k]):
                return False
        return True
    backtracking(n, m, check, display)



#permutation(4, 4)

def combination(n, m):
    def check(a, k):
        for i in range(k):
            if(a[i] >= a[k]):
                return False
        return True
    backtracking(n, m, check, display)

#combination(6, 3)
    
def nqueen(n):
    def check(a, k):
        for i in range(k):
            if(a[i] == a[k] or abs(a[i] - a[k]) == k - i):
                return False
        return True
    backtracking(n, n, check, lambda a : print(a))


#nqueen(4)



def deepFirstSearch(g):
    def dfs(n, check, handle):
        k, a = 0, [-1 for i in range(n)]
        while(k >= 0):
            a[k] += 1
            while(a[k] < n and not check(a, k)):
                a[k] += 1
            if(a[k] == n):                
                k -= 1
            else:
                if(k == n - 1):
                    handle(a)
                else:
                    k += 1
                    a[k] = -1
    def check(a, k):
        if(k == 0):
            return True
        for i in range(k):
            if(a[i] == a[k]):
                return False
        for i in range(k-1, -1, -1):            
            if(g[a[i]][a[k]] == 1):
                print(i, k)
                return True           
        return False    
    n = len(g)
    dfs(n, check, display)
"""
mat = [
    [0, 1, 1, 0, 0],
    [1, 0, 0, 1, 1],
    [1, 0, 0, 0, 1],
    [0, 1, 0, 0, 1],
    [0, 1, 1, 1, 0]
]
"""
mat = [
    [0, 1, 1, 1, 0],
    [1, 0, 0, 0, 1],
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
    [0, 1, 0, 0, 0]
]
deepFirstSearch(mat)







    

File: codes/ruby/algorithm/backtracking.rb

module Backtracking
  
  ALPHABET = Array('A'..'Z')
  DISPLAY = lambda {|a| p a.map{|i|ALPHABET[i]}}
  
  
  def backtrace(n, m, handle = DISPLAY)
    k, a = 0, Array.new(m, -1)
    while k >= 0
      a[k] += 1
      while a[k] < n && !yield(a, k)
        a[k] += 1
      end
      if a[k] == n || k == m
        k -= 1
      else
        if k == m - 1
          handle.call(a)
        else
          k += 1
          a[k] = -1
        end
      end
    end
  end
  
  def backtrack(n, m, handle = DISPLAY)
    dfs = lambda{|a, k|
      #for i in (0...n) do
      (0...n).each do |i|
        a[k] = i
        if yield(a, k)
          if k == m - 1
            handle.call(a)
          else
            dfs.call(a, k + 1)
          end
        end
      end
    }
    a = Array.new(m, -1)
    dfs.call(a, 0)
  end

  
  def counter(n, m, &handle)
    handle = DISPLAY unless handle
    backtrace(n, m, handle) do |a, k|
      true
    end
  end
  
  def permutation(n, m, &handle)
    handle = DISPLAY unless handle
    backtrack(n, m, handle) do |a, k|
      succ = true
      for i in (0...k) do 
       if a[i] == a[k]
         succ = false
         break
       end
      end
      succ
    end
  end
  
  def combination(n, m, &handle)
    handle = DISPLAY unless handle
    backtrace(n, m, handle) do |a, k|
      succ = true
      (0...k).each do |i|
        if a[i] >= a[k]
          succ = false
          break
        end
      end
      succ
    end
  end

  def nqueen(n, &handle)
    backtrace(n, n, handle) do |a, k|
      succ = true
      (0...k).each do |i|
        if a[i] == a[k] || (a[i] - a[k]).abs == k - i
          succ = false
          break
        end
      end
      succ
    end
  end
    
  def permutation2(a, n)
    if n == 0
      p a
    else
      n.times do |i|
        a[n - 1], a[i] = a[i], a[n - 1]
        permutation2(a, n - 1)
        a[n - 1], a[i] = a[i], a[n - 1]
      end
    end
  end
  
end

if __FILE__ == $0
  include Backtracking
  begin
    puts 'counter =>'
    counter(3, 3)
  
    puts 'permutation =>'
    permutation(3, 3){|a|p a.map{|i| i + 1}}
  
    puts 'combination =>'
    combination(5, 2){|a|p a.map{|i| i + 1}}
  
    puts 'nqueue => '
    nqueen(4){|a|p a.map{|i| i + 1}}
  
    puts 'permutation2 =>'
    permutation2([1, 2, 3], 3)
  
  rescue Exception => e
    puts e
    puts e.backtrace
  end
end